File: app/lib/services/rule_engine.dart
//lib/services/rule_engine.dart
import '../models/campaign.dart';
import '../models/rule.dart';
import 'notification_service.dart';

class RuleEngine {
  final NotificationService _notifications;

  RuleEngine(this._notifications);

  // ═══ EVALUATE ALL RULES ═══
  List<RuleAction> evaluate(List<AutoRule> rules, List<Campaign> campaigns) {
    final actions = <RuleAction>[];

    for (final rule in rules) {
      if (!rule.enabled) continue;

      for (final campaign in campaigns) {
        if (!campaign.isActive && rule.actionType != 'alert') continue;

        final metricValue = _getMetricValue(rule.metric, campaign);
        if (metricValue == null) continue;

        final triggered = _compare(metricValue, rule.operator, rule.threshold);

        if (triggered) {
          actions.add(RuleAction(
            rule: rule,
            campaign: campaign,
            metricValue: metricValue,
            action: _buildAction(rule, campaign),
          ));
        }
      }
    }

    return actions;
  }

  // ═══ EXECUTE ACTIONS ═══
  Future<List<RuleResult>> executeActions(List<RuleAction> actions) async {
    final results = <RuleResult>[];

    for (final action in actions) {
      try {
        final result = await _executeAction(action);
        results.add(result);

        await _notifications.showAutoPilotAction(
          title: action.rule.name,
          body: result.description,
          ruleId: action.rule.id,
          campaignId: action.campaign.id,
        );
      } catch (e) {
        results.add(RuleResult(
          action: action,
          success: false,
          description: 'Failed: ${e.toString()}',
        ));
      }
    }

    return results;
  }

  // ═══ GET METRIC VALUE ═══
  double? _getMetricValue(String metric, Campaign campaign) {
    return switch (metric) {
      'roas'        => campaign.roas,
      'cpa'         => campaign.cpa,
      'ctr'         => campaign.ctr,
      'cpc'         => campaign.cpc,
      'cpm'         => campaign.cpm,
      'frequency'   => campaign.frequency,
      'spend'       => campaign.spend,
      'budget_util' => campaign.dailyBudget > 0
          ? (campaign.spend / campaign.dailyBudget) * 100
          : null,
      'impressions' => campaign.impressions.toDouble(),
      'clicks'      => campaign.clicks.toDouble(),
      'conversions' => campaign.conversions.toDouble(),
      _             => null,
    };
  }

  // ═══ COMPARE VALUES ═══
  bool _compare(double value, String operator, double threshold) {
    return switch (operator) {
      '<'  => value < threshold,
      '>'  => value > threshold,
      '<=' => value <= threshold,
      '>=' => value >= threshold,
      '==' => value == threshold,
      '!=' => value != threshold,
      _    => false,
    };
  }

  // ═══ BUILD ACTION DESCRIPTION ═══
  ActionDetails _buildAction(AutoRule rule, Campaign campaign) {
    // Parse actionValue String? → double safely
    final pct = double.tryParse(rule.actionValue ?? '') ?? 0.0;

    return switch (rule.actionType) {
      'pause' => ActionDetails(
          type: 'pause',
          description:
              'Pause "${campaign.name}" — ${rule.metric} ${rule.operator} ${rule.threshold}',
          newStatus: 'PAUSED',
        ),
      'scale_budget' => ActionDetails(
          type: 'scale_budget',
          description:
              'Scale "${campaign.name}" budget +${pct.toInt()}%',
          newBudget: campaign.dailyBudget * (1 + pct / 100),
        ),
      'reduce_budget' => ActionDetails(
          type: 'reduce_budget',
          description:
              'Reduce "${campaign.name}" budget -${pct.toInt()}%',
          newBudget: campaign.dailyBudget * (1 - pct / 100),
        ),
      'alert' => ActionDetails(
          type: 'alert',
          description:
              'Alert: "${campaign.name}" ${rule.metric} is ${_getMetricValue(rule.metric, campaign)?.toStringAsFixed(1)}',
        ),
      'alert_and_pause' => ActionDetails(
          type: 'alert_and_pause',
          description:
              'Alert + Pause: "${campaign.name}" — ${rule.metric} exceeded threshold',
          newStatus: 'PAUSED',
        ),
      _ => ActionDetails(
          type: 'unknown',
          description: 'Unknown action',
        ),
    };
  }

  // ═══ EXECUTE SINGLE ACTION ═══
  Future<RuleResult> _executeAction(RuleAction action) async {
    await Future.delayed(const Duration(milliseconds: 200));
    return RuleResult(
      action: action,
      success: true,
      description: action.action.description,
    );
  }
}

// ═══ DATA CLASSES ═══
class RuleAction {
  final AutoRule rule;
  final Campaign campaign;
  final double metricValue;
  final ActionDetails action;

  const RuleAction({
    required this.rule,
    required this.campaign,
    required this.metricValue,
    required this.action,
  });
}

class ActionDetails {
  final String type;
  final String description;
  final String? newStatus;
  final double? newBudget;

  const ActionDetails({
    required this.type,
    required this.description,
    this.newStatus,
    this.newBudget,
  });
}

class RuleResult {
  final RuleAction action;
  final bool success;
  final String description;

  const RuleResult({
    required this.action,
    required this.success,
    required this.description,
  });
}

File: worker/src/services/rule-engine.ts
import type { Bindings, AutoRule, CampaignParsed } from '../types';
import * as MetaApi from './meta-api';
import { notify } from './fcm';

// ─── Types ───
type CompoundRule = AutoRule & {
  condition2_metric?: string;
  condition2_operator?: string;
  condition2_threshold?: number;
};

// ─── Minimum spend guard ───
// Never evaluate rules on campaigns with less than this spend
// Protects learning phase campaigns
const MIN_SPEND_FOR_EVALUATION = 500; // ₹500

// ─── Extract metric value from campaign ───
function extractMetric(c: CampaignParsed, metric: string): number {
  const map: Record<string, number> = {
    roas:         c.roas,
    cpa:          c.cpa,
    ctr:          c.ctr,
    cpc:          c.cpc,
    cpm:          c.cpm,
    frequency:    c.frequency,
    spend:        c.spend,
    impressions:  c.impressions,
    clicks:       c.clicks,
    conversions:  c.conversions,
    leads:        c.leads ?? 0,
    budget_util:  c.daily_budget > 0
      ? (c.spend / c.daily_budget) * 100
      : 0,
  };
  return map[metric] ?? 0;
}

// ─── Evaluate single condition ───
function check(value: number, op: string, threshold: number): boolean {
  switch (op) {
    case '<':  return value < threshold;
    case '>':  return value > threshold;
    case '<=': return value <= threshold;
    case '>=': return value >= threshold;
    case '==': return value === threshold;
    case '!=': return value !== threshold;
    default:   return false;
  }
}

// ─── Evaluate compound rule (condition1 AND condition2 if present) ───
function evaluateRule(
  rule: CompoundRule,
  campaign: CampaignParsed,
): boolean {
  // Primary condition
  const v1 = extractMetric(campaign, rule.metric);
  if (!check(v1, rule.operator, rule.threshold)) return false;

  // Secondary condition (AND logic) — only if defined
  if (
    rule.condition2_metric &&
    rule.condition2_operator &&
    rule.condition2_threshold != null
  ) {
    const v2 = extractMetric(campaign, rule.condition2_metric);
    if (!check(v2, rule.condition2_operator, rule.condition2_threshold)) {
      return false;
    }
  }

  return true;
}

// ─── Built-in safety guards ───
// Applied regardless of rule config — protects learning phase
function passesGuards(
  campaign: CampaignParsed,
  rule: CompoundRule,
): { pass: boolean; reason: string } {

  // Guard 1: Never touch campaigns with < ₹500 spend (learning phase)
  if (campaign.spend < MIN_SPEND_FOR_EVALUATION) {
    return {
      pass: false,
      reason: `Skipped — spend ₹${campaign.spend} below minimum ₹${MIN_SPEND_FOR_EVALUATION} (learning phase)`,
    };
  }

  // Guard 2: Never pause a campaign that's already paused
  if (
    (rule.action_type === 'pause' || rule.action_type === 'alert_and_pause') &&
    campaign.status !== 'ACTIVE'
  ) {
    return {
      pass: false,
      reason: `Skipped — campaign already ${campaign.status}`,
    };
  }

  // Guard 3: Never scale a fatigued campaign (frequency > 3.5)
  if (rule.action_type === 'scale_budget' && campaign.frequency > 3.5) {
    return {
      pass: false,
      reason: `Skipped scale — frequency ${campaign.frequency.toFixed(2)}x is too high (fatigue risk)`,
    };
  }

  // Guard 4: Never reduce budget below ₹200/day
  if (rule.action_type === 'reduce_budget') {
    const pct = Number(rule.action_value ?? 30);
    const newBudget = campaign.daily_budget * (1 - pct / 100);
    if (newBudget < 200) {
      return {
        pass: false,
        reason: `Skipped — budget would drop below ₹200/day minimum`,
      };
    }
  }

  return { pass: true, reason: '' };
}

// ─── Main: Evaluate all active rules against all campaigns ───
export async function evaluateRules(env: Bindings): Promise<number> {

  // 1. Fetch active rules from D1
  const rulesResult = await env.DB.prepare(
    `SELECT id, name, metric, operator, threshold,
            action_type, action_value, enabled,
            condition2_metric, condition2_operator, condition2_threshold
     FROM rules
     WHERE enabled = 1`,
  ).all<CompoundRule>();

  const rules = rulesResult.results ?? [];
  if (!rules.length) {
    console.log('[Rules] No active rules found.');
    return 0;
  }

  // 2. Fetch campaigns from Meta (last 7 days for freshest data)
  let campaigns: CampaignParsed[];
  try {
    const raw = await MetaApi.getCampaigns(env, 'last_7d', 50);
    campaigns = raw.map((mc) => {
      const p = MetaApi.parseInsights(mc.insights?.data ?? []) as any;
      return {
        id:               mc.id,
        name:             mc.name,
        objective:        mc.objective,
        status:           mc.effective_status ?? mc.status,
        daily_budget:     parseInt(mc.daily_budget ?? '0') / 100,
        lifetime_budget:  parseInt(mc.lifetime_budget ?? '0') / 100,
        ...p,
      } as CampaignParsed;
    });
  } catch (err) {
    console.error('[Rules] Failed to fetch campaigns:', err);
    return 0;
  }

  console.log(`[Rules] Evaluating ${rules.length} rules × ${campaigns.length} campaigns`);

  let triggered = 0;
  let skipped = 0;

  for (const rule of rules) {
    for (const campaign of campaigns) {

      // Safety guards first
      const guard = passesGuards(campaign, rule);
      if (!guard.pass) {
        skipped++;
        console.log(`[Rules] "${rule.name}" → "${campaign.name}": ${guard.reason}`);
        continue;
      }

      // Compound condition evaluation
      if (!evaluateRule(rule, campaign)) continue;

      // ── Triggered ──
      triggered++;
      let desc = '';

      try {
        switch (rule.action_type) {

          case 'pause':
            await MetaApi.updateCampaignStatus(env, campaign.id, 'PAUSED');
            desc = `Paused "${campaign.name}" — ${rule.metric} was ${extractMetric(campaign, rule.metric).toFixed(2)}`;
            if (rule.condition2_metric) {
              const v2 = extractMetric(campaign, rule.condition2_metric);
              desc += ` AND ${rule.condition2_metric} was ${v2.toFixed(2)}`;
            }
            break;

          case 'scale_budget': {
            const pct = Number(rule.action_value ?? 20);
            const newBudget = campaign.daily_budget * (1 + pct / 100);
            await MetaApi.updateCampaignBudget(env, campaign.id, newBudget);
            desc = `Scaled "${campaign.name}" +${pct}% → ₹${Math.round(newBudget)}/day (ROAS: ${campaign.roas.toFixed(2)}x)`;
            break;
          }

          case 'reduce_budget': {
            const pct = Number(rule.action_value ?? 30);
            const newBudget = Math.max(
              campaign.daily_budget * (1 - pct / 100),
              200,
            );
            await MetaApi.updateCampaignBudget(env, campaign.id, newBudget);
            desc = `Reduced "${campaign.name}" -${pct}% → ₹${Math.round(newBudget)}/day`;
            break;
          }

          case 'alert':
            desc = `Alert: "${campaign.name}" ${rule.metric} = ${extractMetric(campaign, rule.metric).toFixed(2)} (threshold: ${rule.threshold})`;
            break;

          case 'alert_and_pause':
            await MetaApi.updateCampaignStatus(env, campaign.id, 'PAUSED');
            desc = `Alert+Paused: "${campaign.name}" — ${rule.metric} = ${extractMetric(campaign, rule.metric).toFixed(2)}`;
            break;

          default:
            desc = `Unknown action "${rule.action_type}" for rule "${rule.name}"`;
        }
      } catch (err) {
        console.error(`[Rules] Action failed — "${rule.name}" on "${campaign.name}":`, err);
        desc = `Action failed: ${(err as Error).message}`;
      }

      // Update trigger count in D1
      await env.DB.prepare(
        `UPDATE rules
         SET triggered_count = triggered_count + 1,
             last_triggered  = datetime('now')
         WHERE id = ?`,
      ).bind(rule.id).run();

      // Log to activity_log
      await env.DB.prepare(
        `INSERT INTO activity_log
           (id, type, title, description, campaign_id, rule_id, created_at)
         VALUES (?, ?, ?, ?, ?, ?, datetime('now'))`,
      ).bind(
        crypto.randomUUID(),
        'rule_triggered',
        rule.name,
        desc,
        campaign.id,
        rule.id,
      ).run();

      // FCM push
      await notify(env, 'rule', `⚡ ${rule.name}`, desc, {
        campaign_id: campaign.id,
        rule_id:     String(rule.id),
        action_type: rule.action_type,
      });

      console.log(`[Rules] TRIGGERED: "${rule.name}" → ${desc}`);
    }
  }

  // Invalidate campaign cache if anything changed
  if (triggered > 0) {
    await Promise.allSettled([
      env.CACHE.delete('campaigns:last_7d:50'),
      env.CACHE.delete('campaigns:last_30d:50'),
      env.CACHE.delete('campaigns:last_14d:50'),
    ]);
  }

  console.log(`[Rules] Done: ${triggered} triggered, ${skipped} skipped by guards`);
  return triggered;
}